File: api/src/main/java/com/delivery/common/dto/CommonMeta.java
package com.delivery.common.dto;

import com.delivery.common.status.ErrorStatusCode;
import com.delivery.common.status.SuccessStatusCode;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class CommonMeta {
    private Integer code;

    private String message;

    private String description;

    public static CommonMeta ok() {
        return CommonMeta.builder()
                .code(SuccessStatusCode.OK.getCode())
                .message(SuccessStatusCode.OK.getMessage())
                .description("성공")
                .build();
    }

    public static CommonMeta error(ErrorStatusCode errorStatusCode) {
        return CommonMeta.builder()
                .code(errorStatusCode.getCode())
                .message(errorStatusCode.getMessage())
                .build();
    }

    public static CommonMeta error(ErrorStatusCode errorStatusCode, Throwable tx) {
        return CommonMeta.builder()
                .code(errorStatusCode.getCode())
                .message(errorStatusCode.getMessage())
                .description(tx.getLocalizedMessage())
                .build();
    }

    public static CommonMeta error(ErrorStatusCode errorStatusCode, String description) {
        return CommonMeta.builder()
                .code(errorStatusCode.getCode())
                .message(errorStatusCode.getMessage())
                .description(description)
                .build();
    }
}

File: common/src/main/kotlin/com/delivery/common/dto/CommonMeta.kt
package com.delivery.common.dto

import com.delivery.common.status.StatusCode
import com.delivery.common.status.SuccessStatusCode

class CommonMeta(
    private val code: Int,
    private val message: String,
    private val description: String? = null,
) {
    companion object {
        fun ok(): CommonMeta {
            return CommonMeta(
                SuccessStatusCode.OK.getCode(),
                SuccessStatusCode.OK.getMessage(),
                "성공",
            )
        }

        fun error(errorStatusCode: StatusCode): CommonMeta {
            return CommonMeta(
                errorStatusCode.getCode(),
                errorStatusCode.getMessage(),
            )
        }

        fun error(
            errorStatusCode: StatusCode,
            t: Throwable,
        ): CommonMeta {
            return CommonMeta(
                errorStatusCode.getCode(),
                errorStatusCode.getMessage(),
                description = t.localizedMessage
            )
        }

        fun error(
            errorStatusCode: StatusCode,
            description: String,
        ): CommonMeta {
            return CommonMeta(
                errorStatusCode.getCode(),
                errorStatusCode.getMessage(),
                description = description
            )
        }
    }
}